### PR TITLE
Add group percentages to oncoprint

### PR DESCRIPTION
**Pull request by alisman**
- [x] Show group percentages in Oncoprint when gaps are turned on
- [x] Show labels in SVG download
- [x] Should also work in Oncoprinter
- [x] Should work in merged tracks

### CODE REVIEW

File: packages/oncoprintjs/src/js/oncoprintwebglcellview.ts [hunk 1]
```diff
@@ -1898,6 +2065,34 @@ export default class OncoprintWebGLCellView {
...
+                        });
+
+                        textElt.textContent = gap.labelFormatter();
+                        root.appendChild(textElt);
+                    });
+                }
+            }
+
+            // aaron
+            // aaron
```
Comment (onursumer) [inline packages/oncoprintjs/src/js/oncoprintwebglcellview.ts:2094]: we should remove/replace these comments

File: src/pages/staticPages/tools/oncoprinter/OncoprinterGeneticUtils.ts [hunk 2]
```diff
@@ -608,17 +618,49 @@ export function getGeneticTracks(
...
+                    model.data_groups.update(model);
+                    const groupsByTrackMap = model.data_groups.get();
+
+                    const percentagesForGroups = groupsByTrackMap[gene][0].map(
+                        groupData => {
+                            return getPercentAltered(
+                                groupData as OncoprinterGeneticTrackDatum[]
+                            );
+                        }
+                    );
```
Comment (onursumer) [inline src/pages/staticPages/tools/oncoprinter/OncoprinterGeneticUtils.ts:649]: would be simpler if we could do something like

```suggestion
                    const percentagesForGroups = groupsByTrackMap[gene][0].map(getPercentAltered);
```

File: src/shared/components/oncoprint/OncoprintUtils.ts [hunk 3]
```diff
@@ -804,6 +802,95 @@ export function makeGeneticTrackWith({
...
+                                                    geneSymbolArray,
+                                                    gene =>
+                                                        sequencedPatientKeysByGene[
+                                                            gene
+                                                        ]
+                                                ),
+                                                groupData.map((d: any) => d.uid)
+                                            );
+                                        })
+                                        .value();
```
Comment (onursumer) [inline src/shared/components/oncoprint/OncoprintUtils.ts:867]: this `else` block is almost identical to the `if` block above except the variables `sequencedSampleKeysByGene` vs `sequencedPatientKeysByGene`. probably better to extract the logic to a function and reuse.

File: packages/oncoprintjs/src/js/oncoprintwebglcellview.ts [hunk 4]
```diff
@@ -294,9 +308,43 @@ export default class OncoprintWebGLCellView {
...
+                            $(
+                                `<span>${overlappingGap.data.tooltipFormatter()}</span>`
+                            ),
+                            false
+                        );
+                    }
+
+                    if (!overlapping_data && !overlappingGap) {
+                        tooltip.hideIfNotAlreadyGoingTo(150);
+                    }
```
Comment (onursumer) [inline packages/oncoprintjs/src/js/oncoprintwebglcellview.ts:347]: so much work to show a tooltip but I guess we have no other choice :disappointed:

File: packages/oncoprintjs/src/js/oncoprintwebglcellview.ts [hunk 5]
```diff
@@ -364,6 +427,27 @@ export default class OncoprintWebGLCellView {
...
+            //ctx.clearColor(1.0, 1.0, 1.0, 1.0);
+            //ctx.clear(ctx.COLOR_BUFFER_BIT | ctx.DEPTH_BUFFER_BIT);
+            // ctx.viewportWidth = canvas.width;
+            // ctx.viewportHeight = canvas.height;
+            // ctx.viewport(0, 0, ctx.viewportWidth, ctx.viewportHeight);
+            //ctx.enable(ctx.DEPTH_TEST);
+            //ctx.enable(ctx.BLEND);
+            //ctx.blendEquation(ctx.FUNC_ADD);
+            //ctx.blendFunc(ctx.SRC_ALPHA, ctx.ONE_MINUS_SRC_ALPHA);
+            //ctx.depthMask(false);
```
Comment (onursumer) [inline packages/oncoprintjs/src/js/oncoprintwebglcellview.ts:443]: should we remove this or add a comment why we commented it out?

File: packages/oncoprintjs/src/js/oncoprintwebglcellview.ts [hunk 6]
```diff
@@ -534,11 +618,34 @@ export default class OncoprintWebGLCellView {
...
+        //ctx.clearColor(1.0, 1.0, 1.0, 1.0);
+        //ctx.clear(ctx.COLOR_BUFFER_BIT | ctx.DEPTH_BUFFER_BIT);
+        //ctx. = canvas.width;
+        //ctx.viewportHeight = canvas.height;
+        //ctx.viewport(0, 0, ctx.viewportWidth, ctx.viewportHeight);
+        //ctx.enable(ctx.DEPTH_TEST);
+        //ctx.enable(ctx.BLEND);
+        //ctx.blendEquation(ctx.FUNC_ADD);
+        //ctx.blendFunc(ctx.SRC_ALPHA, ctx.ONE_MINUS_SRC_ALPHA);
+        //ctx.depthMask(false);
```
Comment (onursumer) [inline packages/oncoprintjs/src/js/oncoprintwebglcellview.ts:643]: same as above, except here we have a syntax error `ctx. = canvas.width;` :)

File: packages/oncoprintjs/src/js/oncoprintwebglcellview.ts [hunk 7]
```diff
@@ -1898,6 +2065,34 @@ export default class OncoprintWebGLCellView {
                 track_id,
                 false
             );
+
+            const custom = model.getTrackCustomOptions(track_id);
+            if (gapOffsets[0]) {
+                const gaps = _.isEmpty(model.ids_after_a_gap.get())
+                    ? undefined
+                    : custom.find(t => !!t.gapLabelsFn)?.gapLabelsFn(model);
```
Comment (onursumer) [inline packages/oncoprintjs/src/js/oncoprintwebglcellview.ts:2073]: duplicate code segment. should we define a `getGaps` function? we can probably include this one as well `const custom = model.getTrackCustomOptions(track_id);`

**Review by onursumer — COMMENTED** [review]
This is great. Thanks, Aaron! :1st_place_medal:

Just added a few minor suggestions.

File: packages/oncoprintjs/src/js/oncoprintwebglcellview.ts (hunk 4 above)
Comment (alisman) [reply packages/oncoprintjs/src/js/oncoprintwebglcellview.ts:347]: @onursumer  I think It's been said the hardest problems in computer science are 1) knowing when to clear a cache and 2) making a tooltip over an HTML canvas.

File: packages/oncoprintjs/src/js/oncoprintwebglcellview.ts (hunk 1 above)
Comment (alisman) [reply packages/oncoprintjs/src/js/oncoprintwebglcellview.ts:2094]: why?

File: packages/oncoprintjs/src/js/oncoprintwebglcellview.ts (hunk 7 above)
Comment (alisman) [reply packages/oncoprintjs/src/js/oncoprintwebglcellview.ts:2040]: good idea

**Pull request merged**